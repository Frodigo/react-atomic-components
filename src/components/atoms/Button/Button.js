import React from 'react';
import PropTypes from 'prop-types';

import classes from './Button.module.css';

/**
 * This is the button.
 * @param {*} props
 */
const Button = (props) => {
    const { variant = 'primary', children, onClick, ...rest } = props;

    let buttonClass = null;

    switch (variant) {
        case 'primary': {
            buttonClass = classes.primary;

            break;
        }

        case 'secondary': {
            buttonClass = classes.secondary;

            break;
        }

        case 'tiertary': {
            buttonClass = classes.tiertary;

            break;
        }

        default: {
            buttonClass = '';
        }
    }

    return (
        <button
            type="button"
            className={buttonClass}
            onClick={onClick}
            {...rest}
        >
            {children}
        </button>
    );
};

Button.propTypes = {
    active: PropTypes.oneOf(['true', 'false']),
    children: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    onClick: PropTypes.func,
    variant: PropTypes.oneOf(['primary', 'secondary'])
};

Button.defaultProps = {
    variant: 'primary'
};

export default Button;
