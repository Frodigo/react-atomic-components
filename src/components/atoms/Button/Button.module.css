.button {
  border: 2px solid transparent;
  border-radius: 2px;
  cursor: pointer;
  display: flex;
  justify-content: center;
  font-size: 16px;
  font-weight: var(--font-weight-bold);
  padding: 20px 70px;
  text-transform: uppercase;
  text-align: center;
  text-decoration: none;
  width: 100%;
}

.button[disabled],
.button.disabled {
  opacity: 0.5;
}

.primary {
  composes: button;
  background-color: var(--color-primary);
  color: var(--color-white);
}

.primary.active,
.primary[active='true'] {
  background-color: var(--color-tiertary);
}

.secondary {
  composes: button;
  background-color: var(--color-white);
  border-color: var(--color-primary);
  color: var(--color-primary);
}

.secondary.active,
.secondary[active='true'] {
  background-color: var(--color-primary);
  color: var(--color-white);
}

.tiertary {
  composes: button;
  background-color: transparent;
  border-color: var(--color-white);
  color: var(--color-white);
}

.tiertary.active,
.tiertary[active='true'] {
  background-color: rgba(255, 255, 255, 0.1);
}

.inverse {
  color: var(--color-white);
}
.inverse[disabled],
.inverse.disabled {
  opacity: 0.7;
}

.inverse:focus {
  font-weight: 600;
}

@media (min-width: 768px) {
  .button {
    display: inline-flex;
    width: auto;
  }
}

@media (min-width: 1300px) {
  .button {
    transition: all 0.3s ease-in-out;
  }

  .primary:hover,
  .primary:focus {
    background-color: var(--color-secondary);
  }

  .secondary:hover,
  .secondary:focus {
    background-color: var(--color-primary);
    color: var(--color-white);
  }
}
