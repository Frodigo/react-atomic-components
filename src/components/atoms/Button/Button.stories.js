import React from 'react';

import Button from './Button';

export default {
    title: 'Atoms/Buttons/Button',
    component: Button,
    argTypes: {
        onClick: { action: 'clicked' },
        variant: {
            control: {
                type: 'select',
                options: ['primary', 'secondary']
            }
        }
    }
};

const Template = (args) => <Button {...args} />;

export const PrimaryButton = Template.bind({});
PrimaryButton.args = {
    variant: 'primary',
    children: 'Primary button',
    disabled: false,
    active: false
};

export const SecondaryButton = Template.bind({});
SecondaryButton.args = {
    variant: 'secondary',
    children: 'Secondary button',
    disabled: false,
    active: false
};
