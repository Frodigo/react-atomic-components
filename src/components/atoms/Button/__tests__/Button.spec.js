import React from 'react';
import { render } from '@testing-library/react';

import Button from '../Button';

describe('Button should be rendered', () => {
    const text = 'Call to action';
    const onCLickHandler = jest.fn();

    test('Button renders correctly', () => {
        const { container } = render(
            <Button variant={'primary'} onClick={onCLickHandler}>
                {text}
            </Button>
        );

        expect(container).toMatchSnapshot();
    });

    test('Disabled button renders correctly', () => {
        const { container } = render(
            <Button variant={'secondary'} onClick={onCLickHandler} disabled>
                {text}
            </Button>
        );

        expect(container).toMatchSnapshot();
        expect(container.firstChild.disabled).toBe(true);
    });
});
