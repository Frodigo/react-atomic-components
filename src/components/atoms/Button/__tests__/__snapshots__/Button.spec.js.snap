// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should be rendered Button renders correctly 1`] = `
<div>
  <button
    class="primary"
    type="button"
  >
    Call to action
  </button>
</div>
`;

exports[`Button should be rendered Disabled button renders correctly 1`] = `
<div>
  <button
    class="secondary"
    disabled=""
    type="button"
  >
    Call to action
  </button>
</div>
`;
